### ISSUE TITLE
Level monitoring project

### ISSUE THREAD

**Issue by carlossd77** [wiwidnadw/Visual-basic-net-WiFi-Communication-New-Style#1]
Hello, I'll put here my code of VB and Arduino Ide, the problem I have is that I can't get data from Esp32 to VB but the value is reading properly on Esp32 because I can see it on Serial Monitor:

*************************************************************************************************************************************
**Visual Studio Code:**

    Imports System.Net
    Imports System.Text.Encoding
    Imports System.IO
    Imports System.Net.Sockets
    Public Class Form1

    Dim UDPreceiver As New Sockets.UdpClient(2000)
    Dim UDPsender As New Sockets.UdpClient(0)

    Dim client As TcpClient
    Dim Rx As StreamReader
    Dim Tx As StreamWriter
    Dim rawdata As String
    Dim dataseparate As Array

    Dim nivel_tinaco_barra As Integer

    Private Sub Form1_Load(sender As Object, e As EventArgs) Handles MyBase.Load

    CheckForIllegalCrossThreadCalls = False
    intentar_conexion()

    End Sub

    Function intentar_conexion()
        Try
            client = New TcpClient(TextBoxIP.Text, "2000")
            If client.GetStream.CanRead = True Or client.GetStream.CanWrite Then
                Rx = New StreamReader(client.GetStream)
                Threading.ThreadPool.QueueUserWorkItem(AddressOf Conectado)
            End If
        Catch ex As Exception
            MsgBox("failed to connect")
        End Try
        Return True
    End Function

    Function Conectado()
        If Rx.BaseStream.CanRead = True Then
            Try
                While Rx.BaseStream.CanRead = True
                    rawdata = Rx.ReadLine
                    RichTextBoxReceive.Text = rawdata
                    NivelDeAgua_Tinaco()
                End While
            Catch ex As Exception
                client.Close()
            End Try
            Rx.DiscardBufferedData()
        End If
        Return True
    End Function

***********************************************************************************************************************************
**Arduino Code:**

    #include <NTPClient.h>
    #include <WiFi.h>
    #include <WiFiUdp.h>
    #include <Adafruit_MCP23X17.h>

    #define ON_Board_LED  8
    #define ON_Board_LED2 9
    #define ON_Board_LED3 10
    #define ON_Board_LED4 11
    #define ON_Board_LED5 12
    #define ON_Board_LED6 13
    #define ON_Board_LED7 14
    #define ON_Board_LED8 15

    #define ECHOPIN 25
    #define TRIGPIN 4

    const char* ssid = "INFINITUM229E_2.4";
    const char* password = "ArxR77p8eQ";
    WiFiServer server(2000);
    #define CS_PIN_MCP 5
    Adafruit_MCP23X17 mcp;

    unsigned int localPort = 2000;
    IPAddress DestinationIP(192, 168, 1, 137); 
    WiFiUDP udp;
    NTPClient timeClient(udp);

    char packetBuffer[50];
    String strCon;
    String inString;
    char replyBuffer[50];
    float duration;

    void setup() {
      Serial.begin(115200);
      if (!mcp.begin_SPI(CS_PIN_MCP))
      {
        Serial.println("Error.");
        while (1);
      }
      else {
        Serial.println("Iniciado");
      }
      delay(500);

      WiFi.begin(ssid, password); //--> Connect to your WiFi router
      Serial.println("");

      mcp.pinMode(ON_Board_LED, OUTPUT);
      mcp.pinMode(ON_Board_LED2, OUTPUT);
      mcp.pinMode(ON_Board_LED3, OUTPUT);
      mcp.pinMode(ON_Board_LED4, OUTPUT);
      mcp.pinMode(ON_Board_LED5, OUTPUT);
      mcp.pinMode(ON_Board_LED6, OUTPUT);
      mcp.pinMode(ON_Board_LED7, OUTPUT);
      mcp.pinMode(ON_Board_LED8, OUTPUT);
      pinMode(ECHOPIN, INPUT);
      pinMode(TRIGPIN, OUTPUT);
      pinMode(2, OUTPUT);
      digitalWrite(2, HIGH);

      Serial.print("Connecting");
      while (WiFi.status() != WL_CONNECTED) {
      Serial.print(".");
      digitalWrite(2, LOW);
      delay(250);
      digitalWrite(2, HIGH);
      delay(250);
    
      }
 
      digitalWrite(2, HIGH);  
      Serial.println("");
      Serial.print("Successfully connected to : ");
      Serial.println(ssid);
      Serial.print("NodeMCU IP address : ");
      Serial.println(WiFi.localIP());
      server.begin();
      udp.begin(localPort);
      Serial.print("Local Port : ");
      Serial.println(localPort);
  
      }

      void loop() {
      medicion_nivel();
      }

    void medicion_nivel() {
  
    float distance, nivel;
    WiFiClient client = server.available();
    if (client) {
    Serial.println("New Client.");
  
    while (client.connected()) {
    
    String request = client.readStringUntil('\r');
    Serial.print(request);
    digitalWrite(TRIGPIN, LOW);
    delayMicroseconds(2);
    digitalWrite(TRIGPIN, HIGH);
    delayMicroseconds(20);
    digitalWrite(TRIGPIN, LOW);
    duration = pulseIn(ECHOPIN, HIGH);
    // Determine distance from duration
    // Use 343 metres per second as speed of sound
    // Divide by 1000 as we want millimeters
    distance = (duration / 2) * 0.343;
    nivel = abs((float(distance) / (1320) * 100) - 100);
    Serial.print("Nivel: ");
    Serial.print(nivel);
    Serial.println(" %");
     client.print(nivel);
     }

     delay(100);
     client.stop();
     Serial.println("Client disconnected");     
     }
     }

**Comment by carlossd77** [comment]
Please, I would appreciate if you only focus on "medicion nivel function" because my program has other button functions Im using.
Thank you very much for your help.

**Comment by wiwidnadw** [comment]
from ur explaination, u have issue on your visual studio code and ur arduino code is working properly. So u dont have problem because replace potentiometer (in my project share) into level sensor or whatever sensor u have using there.

when i see ur code on visual studio, u just create function but no one command or condition to execute that function. Thats why visual studio cant receive the value from ESp32, even u have insert connection function on form1_load.

let's see on my project, connection function inserted on button_click events. so when button function is clicked, connection is on. I give suggest, try the same code with my project (open connection with button) is any problem or not? if any just try to change socket from 2000 into 80

**Comment by carlossd77** [comment]
Actually I have Tryied as you said , with button command , but it didn’t work . I would like to share my phone number to be in touch closer +5214272889265 I’d appreciate so much if you what’s app me.

**Comment by wiwidnadw** [comment]
Have u try access ur esp32 from browser ?

**Comment by carlossd77** [comment]
No, but for example, if I put this code on timer, it works properly but the program on VB has a very long lag but makes the measure reading good:

       Private Sub TimerRcv_Tick(sender As Object, e As EventArgs) Handles TimerRcv.Tick
             client = New TcpClient(TextBoxIP.Text, "2000")
             Rx = New StreamReader(client.GetStream)
             rawdata = Rx.ReadLine
             RichTextBoxReceive.Text = rawdata
             NivelDeAgua_Tinaco()
           End Sub


**I'm sorry for the audio, I sent this video to a friend doing my explanation**

https://user-images.githubusercontent.com/124798834/218506452-31765f63-ff2f-45ad-9e50-7670f74e3785.mp4

**Comment by wiwidnadw** [comment]
how many seconds timer's setting? if setting of timer too long u can try to decrease it
but actually, the streamline can work when something event has execute. That's why the code will work when inserted into timer_tick
